File: ruby/difference-of-squares/difference_of_squares.rb
module BookKeeping
  VERSION = 4
end

class Squares
  attr_reader :n

  def initialize(n)
    @n = n
  end

  def square_of_sum
    ((@n * (@n + 1)) / 2) ** 2
  end

  def sum_of_squares
    ((@n * (@n + 1)) * ((2 * @n) + 1)) / 6
  end

  def difference
    square_of_sum - sum_of_squares
  end
end

File: ocaml/difference-of-squares/difference_of_squares.ml
let square_of_sum n =
  let square x = x * x in
  square (n * (n + 1) / 2)

let sum_of_squares n = n * (n + 1) * ((2 * n) + 1) / 6
let difference_of_squares n = square_of_sum n - sum_of_squares n
